Math/NumberTheory/DirichletCharacters.hs

```diff
@@ -0,0 +1,589 @@
+                            , _getModulus'   :: !Natural
+                            }
+              | TwoPTemplate { _getPower2'    :: Int
+                             , _getModulus'   :: !Natural
+                             } -- the modulus is derivable from the other values, but calculation
+                               -- may be expensive, so we pre-calculate it
+                               -- morally getModulus should be a prefactored but seems to be
+                               -- pointless here
+              | TwoTemplate
+
+templateFromCharacter :: DirichletCharacter n -> (Product Natural, [Template])
+templateFromCharacter (Generated t) = traverse go t
+  where go (OddPrime p k g _) = (Product m, OddTemplate p k g m)
+          where p' = unPrime p
+                m = p'^(k-1)*(p'-1)
+        go (TwoPower k _ _) = (Product (2*m), TwoPTemplate k m)
+          where m = bit (k-2)
+        go Two = (Product 1, TwoTemplate)
+
+mkTemplate :: Natural -> (Product Natural, [Template])
+mkTemplate = go . sort . factorise
+  where go :: [(Prime Natural, Word)] -> (Product Natural, [Template])
+        go ((unPrime -> 2, 1): xs) = (Product 1, [TwoTemplate]) <> traverse odds xs
+        go ((unPrime -> 2, wordToInt -> k): xs) = (Product (2*m), [TwoPTemplate k m]) <> traverse odds xs
+          where m = bit (k-2)
+        go xs = traverse odds xs
+        odds :: (Prime Natural, Word) -> (Product Natural, Template)
+        odds (p, k) = (Product m, OddTemplate p k (generator p k) m)
+          where p' = unPrime p
+                m = p'^(k-1)*(p'-1)
+
+-- the validity of the producted dirichletfactor list here requires the template to be valid
+unroll :: [Template] -> Natural -> [DirichletFactor]
+unroll t m = snd (mapAccumL func m t)
+  where func :: Natural -> Template -> (Natural, DirichletFactor)
+        func a (OddTemplate p k g n) = (a1, OddPrime p k g (toRootOfUnity $ a2 % n))
+          where (a1,a2) = quotRem a n
+        func a (TwoPTemplate k n) = (b1, TwoPower k (toRootOfUnity $ a2 % 2) (toRootOfUnity $ b2 % n))
+          where (a1,a2) = quotRem a 2
+                (b1,b2) = quotRem a1 n
+        func a TwoTemplate = (a, Two)
+
+-- | Test if a given Dirichlet character is prinicpal for its modulus: a principal character mod
+-- \(n\) is 1 for \(a\) coprime to \(n\), and 0 otherwise.
+isPrincipal :: DirichletCharacter n -> Bool
+isPrincipal chi = characterNumber chi == 0
+
+-- | Induce a Dirichlet character to a higher modulus. If \(d \mid n\), then \(a \bmod{n}\) can be
+-- reduced to \(a \bmod{d}\). Thus, the multiplicative function on \(\mathbb{Z}/d\mathbb{Z}\)
+-- induces a multiplicative function on \(\mathbb{Z}/n\mathbb{Z}\).
+--
+-- >>> :set -XTypeApplications
+-- >>> chi = indexToChar 5 :: DirichletCharacter 45
+-- >>> chi2 = induced @135 chi
+-- >>> :t chi2
+-- Maybe (DirichletCharacter 135)
+induced :: forall n d. (KnownNat d, KnownNat n) => DirichletCharacter d -> Maybe (DirichletCharacter n)
```

Comment:
If you add `Mod n d ~ 0` to constraints, it would be possible to avoid `Maybe` at all.

Reply:
It would, but it would be harder to work with - even one of the tests would need to prove `Mod (d * n) d ~ 0` (which isn't in `Data.Constraint`). I think it's best to use `Divides d n` from Data.Constraint, and add the lemma `Divides a c => Divides a (b * c)`, but this would belong in a different branch as it's a bit of work.

Reply:
I see. How do you feel about this?
```haskell
induced :: (KnownNat d, KnownNat n) => DirichletCharacter d -> DirichletCharacter (d * n)
```

Reply:
I also tried this; it requires `AllowAmbiguousTypes` in the module, and `TypeApplications` for the user: if `chi :: DirichletCharacter 5` then GHC can't figure out what `d` should be to make `induced chi :: DirichletCharacter 15` work, since `*` is non-injective. The user would instead need to do `induced @3 chi :: DirichletCharacter 15`, at which point we might as well use the version here, as far as I can see.

Reply:
Yeah, I forgot that GHC treats `*` as non-injective because of `0`.